spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/core/PubSubConfiguration.java

```diff
@@ -54,34 +69,74 @@ public Health getHealth() {
+        "Pub/Sub properties have already been initialized; cannot update subscription properties");
+
+    this.subscription = map;
   }
 
-  public Subscriber getSubscriber(String name, String projectId) {
-    ProjectSubscriptionName fullyQualifiedName =
-        PubSubSubscriptionUtils.toProjectSubscriptionName(name, projectId);
-    String fullyQualifiedSubscriptionKey = fullyQualifiedName.toString();
+  public Map<ProjectSubscriptionName, Subscriber> getFullyQualifiedSubscriberProperties() {
+    Assert.notNull(this.fullyQualifiedSubscriptionProperties, "Please call initialize() prior to retrieving properties.");
+    return this.fullyQualifiedSubscriptionProperties;
+  }
 
-    if (this.subscription.containsKey(fullyQualifiedSubscriptionKey)) {
-      return this.subscription.get(fullyQualifiedSubscriptionKey);
-    }
+  /**
+   * Standardizes all subscription properties to be fully qualified. Not thread-safe.
+   *
+   * @param defaultProjectId Project to use with short subscription names
+   */
+  public void initialize(String defaultProjectId) {
+    if (this.fullyQualifiedSubscriptionProperties != null) {
+      logger.warn("Pub/Sub configuration can only be initialized once; ignoring request.");
+      return;
+    }
+
+    Map<ProjectSubscriptionName, Subscriber> fullyQualifiedProps = new HashMap<>();
+    for (Entry<String, Subscriber> entry : this.subscription.entrySet()) {
+      // Subscription name is either a valid short name, or a made-up name with fully-qualified provided as a property
+      Subscriber subProps = entry.getValue();
+      String qualifiedName = subProps.fullyQualifiedName == null ? entry.getKey() : subProps.fullyQualifiedName;
+      ProjectSubscriptionName psn =
+          PubSubSubscriptionUtils.toProjectSubscriptionName(qualifiedName, defaultProjectId);
+      if (fullyQualifiedProps.containsKey(psn)) {
+        logger.warn("Found multiple configurations for " + psn
+            + "; ignoring properties with key " + entry.getKey());
+      } else {
+        fullyQualifiedProps.put(psn, subProps);
+      }
+    }
+
+    this.fullyQualifiedSubscriptionProperties = Collections.unmodifiableMap(fullyQualifiedProps);
+  }
+
+  /**
+   * Returns properties for the specified subscription name and project ID.
+   *
+   * @param name short subscription name
+   * @param projectId subscription project name
+   * @return user-provided subscription properties
+   * @deprecated use {@link #getSubscriptionProperties(ProjectSubscriptionName)} instead.
+   */
+  @Deprecated
+  public Subscriber getSubscriber(String name, String projectId) {
+    return getSubscriptionProperties(PubSubSubscriptionUtils.toProjectSubscriptionName(name, projectId));
+  }
 
-    String subscriptionName = fullyQualifiedName.getSubscription();
-    String projectIdFromFullName = fullyQualifiedName.getProject();
+  public Subscriber getSubscriptionProperties(ProjectSubscriptionName psn) {
```

Comment:
Nice! This really simplifies all the logic. Perhaps we could add a javadoc here?